### PR TITLE
LizmapProject refactor

### PR DESCRIPTION
**Pull request by alagroy-42**
Transforming lizmapProject into a proxy for project + fixing errors on project

### CODE REVIEW

File: lizmap/modules/admin/classes/listProjectDatasource.class.php [hunk 1]
```diff
@@ -22,7 +22,7 @@ public function getData($form)
         $criteria = $form->getData($this->criteriaFrom[0]);
         if ($criteria && array_key_exists($criteria, $this->data)) {
             $rep = lizmap::getRepository($criteria);
-            $projects = $rep->getProjects();
+            $projects = $rep->getProjects(lizmap::getAppContext(), lizmap::getServices());
```
Comment (laurentj) [inline lizmap/modules/admin/classes/listProjectDatasource.class.php:25]: why these parameters? the repository object should already have these references..

File: lizmap/modules/lizmap/classes/lizmapProject.class.php [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+    /**
+     * constructor.
+     *
+     * @param string           $key      : the project name
+     * @param LizmapRepository $rep      : the repository
+     * @param mixed            $context
+     * @param mixed            $services
+     */
+    public function __construct($key, $rep, $context, $services)
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapProject.class.php:36]: wrong signature, it should use the same parameters as the original class. The goal of this class is to not break modules that are still using lizmapProject

File: lizmap/modules/lizmap/classes/lizmapProject.class.php [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+
+/**
+ * @deprecated
+ * @FIXME getXml, getComposer
+ * Verify this methods are not used in external modules so we can delete them without risk, otherwise, we have to implement them
+ * in Project and call it here
+ */
+class lizmapProject
+{
+    /**
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapProject.class.php:23]: `const CACHE_FORMAT_VERSION = 1;` is missing

File: lizmap/modules/lizmap/classes/lizmapProject.class.php [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    {
+        return $this->proj->getLayers();
+    }
+
+    public function getLayer($layerId)
+    {
+        return $this->proj->getLayer($layerId);
+    }
+
+    public function getXmlLayer($layerId)
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapProject.class.php:100]: This methods was not in the original class

File: lizmap/modules/lizmap/classes/lizmapProject.class.php (hunk 3 above)
Comment (laurentj) [reply lizmap/modules/lizmap/classes/lizmapProject.class.php:23]: some methods from QgisProject are also missing:

```
public function getQgisProjectVersion()
public function getRelations()
public function getThemes()
```
Probably others are missing. Check it.

File: lizmap/modules/lizmap/classes/lizmapRepository.class.php [hunk 5]
```diff
@@ -169,14 +169,14 @@ public function update($data, $ini)
         return $modified;
     }
 
-    public function getProject($key)
+    public function getProject($key, $context, $services)
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapRepository.class.php:172]: these parameters:

* are not documented
* should be given to the constructor instead

File: lizmap/modules/lizmap/classes/lizmapRepository.class.php [hunk 6]
```diff
@@ -189,7 +189,7 @@ public function getProject($key)
         return $proj;
     }
 
-    public function getProjects()
+    public function getProjects($context, $services)
```
Comment (laurentj) [inline lizmap/modules/lizmap/classes/lizmapRepository.class.php:192]: same as `getProject()`

File: lizmap/modules/lizmap/lib/Project/ProjectConfig.php [hunk 7]
```diff
@@ -44,20 +44,30 @@ class ProjectConfig
...
+            $this->cfgContent = json_decode($fileContent);
+            if ($this->cfgContent === null) {
                 throw new UnknownLizmapProjectException('The file '.$cfgFile.' cannot be decoded.');
             }
         } else {
             foreach ($data as $prop => $value) {
-                if (array_key_exists($this->cachedProperties, $prop)) {
+                if (in_array($prop, $this->cachedProperties)) {
+                    if ($prop == 'cfgContent') {
+                        $this->{$prop} = json_decode(json_encode($value));
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Project/ProjectConfig.php:67]: why encode/decode ?

File: lizmap/modules/lizmap/lib/Project/ProjectConfig.php [hunk 8]
```diff
@@ -71,24 +81,24 @@ public function __construct($cfgFile, $data = null)
...
-
         foreach ($this->cachedProperties as $prop) {
-            if (!isset($this->{$prop})) {
+            if (!isset($this->{$prop}) || isset($data[$prop])) {
                 continue;
             }
-            if ($prop == 'cfg') {
-                $data['cfg'] = $this->cfg;
+            if ($prop == 'cfgContent') {
+                $data['cfgContent'] = json_decode(json_encode($this->cfgContent), true);
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Project/ProjectConfig.php:101]: why decode/encode?

File: lizmap/modules/lizmap/lib/Project/QgisProject.php [hunk 9]
```diff
@@ -358,23 +363,24 @@ public function getLayerDefinition($layerId)
 
     /**
      * @param $layerId
+     * @param mixed $layers
      *
-     * @return null|\QgisMapLayer|\QgisVectorLayer
+     * @return null|\qgisMapLayer|\qgisVectorLayer
      */
-    public function getLayer($layerId)
+    public function getLayer($layerId, $layers)
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Project/QgisProject.php:370]: why this new parameter?

File: lizmap/modules/lizmap/module.xml [hunk 10]
```diff
@@ -36,6 +36,6 @@
         <class name="qgisLayerDbFieldsInfo" file="classes/qgisLayerDbFieldsInfo.class.php"/>
         <class name="qgisFormControlsInterface" file="classes/qgisFormControlsInterface.php"/>
         <class name="lizmapServices" file="classes/lizmapServices.class.php"/>
-        <namespacePathMap name="Lizmap" dir="lib"/>
+        <namespacePathMap name="Lizmap" dir="lib" suffix=".php"/>
```
Comment (laurentj) [inline lizmap/modules/lizmap/module.xml:39]: it should be useless. Is there an issue when it is missing? if yes, we should probably fix something into Jelix.

File: lizmap/modules/view/zones/ajax_view.zone.php [hunk 11]
```diff
@@ -36,7 +36,7 @@ protected function _prepareTpl()
             if (jAcl2::check('lizmap.repositories.view', $r)) {
                 $lrep = lizmap::getRepository($r);
                 $mrep = new lizmapMainViewItem($r, $lrep->getData('label'));
-                $lprojects = $lrep->getProjects();
+                $lprojects = $lrep->getProjects(lizmap::getAppContext(), lizmap::getServices());
```
Comment (laurentj) [inline lizmap/modules/view/zones/ajax_view.zone.php:39]: This change should be reverted. see above.

File: lizmap/modules/view/zones/main_view.zone.php [hunk 12]
```diff
@@ -45,7 +45,7 @@ protected function _prepareTpl()
             if (jAcl2::check('lizmap.repositories.view', $r)) {
                 $lrep = lizmap::getRepository($r);
                 $mrep = new lizmapMainViewItem($r, $lrep->getData('label'));
-                $lprojects = $lrep->getProjects();
+                $lprojects = $lrep->getProjects(lizmap::getAppContext(), lizmap::getServices());
```
Comment (laurentj) [inline lizmap/modules/view/zones/main_view.zone.php:48]: This change should be reverted. see above.

File: lizmap/modules/lizmap/lib/Project/Project.php [hunk 13]
```diff
@@ -561,10 +621,10 @@ public function hasEditionLayers()
                             // User group(s) correspond to the groups given for this edition layer
                             // or user is admin
                             ++$count;
-                            unset($editionLayers->{$key}->acl);
+                            $this->cfg->unsetProperty('editionLayers->'.$key.'->acl');
```
Comment (laurentj) [inline lizmap/modules/lizmap/lib/Project/Project.php:624]: this method seems to be undefined.

**Review by laurentj — CHANGES_REQUESTED** [review]

File: lizmap/modules/lizmap/classes/lizmapProject.class.php (hunk 4 above)
Comment (alagroy-42) [reply lizmap/modules/lizmap/classes/lizmapProject.class.php:100]: It was in qgisProject

**Review by laurentj — APPROVED** [review]

**Pull request merged**